### PR TITLE
fix: add mapping of event_source instead of static

### PR DESCRIPTION
**Pull request by ItsSudip**
## Summary

Enhance TikTok Ads event source handling: add support for all valid event source values, strict validation, and comprehensive unit tests.

---

### What was changed

- Updated the `getEventSource` utility in `src/v0/destinations/tiktok_ads/util.js` to return only supported event source values (`web`, `app`, `offline`, `crm`), defaulting to `'web'` if neither `eventSource` nor `channel` is valid.
- Added/updated unit tests in `util.test.js` to cover all valid, invalid, and edge cases for event source resolution.
- Integrated the new utility in `transformV2.js` to ensure only valid event sources are sent to TikTok Ads.

---

### Why this change was made

Previously, the event source was always set to the static value `'web'`, regardless of the actual source of the event. This limited flexibility and did not accurately represent the true event source, potentially impacting data granularity and downstream analytics. The update allows for dynamic and validated event source assignment, in line with TikTok Ads requirements.

---

### How does this address the problem

- Enables support for all TikTok-supported event source values, improving data accuracy.
- Prevents invalid event source values from being sent to TikTok Ads.
- Ensures consistent and predictable behavior for event source resolution.
- Improves test coverage and reliability of the utility.

---

### Success criteria

- Only `'web'`, `'app'`, `'offline'`, or `'crm'` are returned as event sources.
- All new and existing tests pass.
- No regressions in related TikTok Ads functionality.

---

### Additional notes

- No breaking changes.
- Change to public documentation is required as we don't have any information regarding event_source.
- Reference: [TikTok Business API Docs](https://business-api.tiktok.com/portal/docs?id=1771100779668482)

## What is the related Linear task?

Resolves INT-3933

@coderabbitai review

<hr>

### Developer checklist

- [x] My code follows the style guidelines of this project

- [x] **No breaking changes are being introduced.**

- [x] All related docs linked with the PR?

- [x] All changes manually tested?

- [x] Any documentation changes needed with this change?

- [x] Is the PR limited to 10 file changes?

- [x] Is the PR limited to one linear task?

- [x] Are relevant unit and component test-cases added in **new readability format**?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Comment by ItsSudip** [comment]
> base branch main?

I did it intentionally. We might need to do a hotfix.

File: src/v0/destinations/tiktok_ads/util.js [hunk 1]
```diff
@@ -60,4 +60,29 @@ const hashUserField = (user) => {
...
+ * https://business-api.tiktok.com/portal/docs?id=1771100779668482
+ * @param {*} message
+ * @returns (string): The resolved, valid event source — one of 'web', 'app', 'offline', or 'crm'. Defaults to 'web' if none are valid.
+ */
+const getEventSource = ({ properties, channel }) => {
+  const supportedSources = ['web', 'app', 'offline', 'crm'];
+
+  const isSupported = (source) => supportedSources.includes(source);
+
+  const { eventSource } = properties || {};
```
Comment (koladilip) [inline src/v0/destinations/tiktok_ads/util.js:75]: we need to refer here rudderstack docs also. I could find where are we documenting that in properties we need to put eventSource?

Comment (ItsSudip) [reply src/v0/destinations/tiktok_ads/util.js:75]: Yes, we have to update our doc, as I mentioned in the pr description. I have already checked with the product regarding this mapping. You can check the Slack thread for more information.

File: src/v0/destinations/tiktok_ads/transformV2.js [hunk 2]
```diff
@@ -683,17 +683,19 @@ const batchEvents = (eventsChunk) => {
   const events = [];
   let data = [];
   let metadata = [];
+  let eventSource = 'web';
   const { destination } = eventsChunk[0];
   const { pixelCode } = destination.Config;
   eventsChunk.forEach((event) => {
     const eventData = event.message[0]?.body.JSON.data;
+    eventSource = event.message[0]?.body.JSON.event_source;
```
[the remark this answers is not in the document]
Comment (ItsSudip) [reply src/v0/destinations/tiktok_ads/transformV2.js:691]: Good catch. We have considered this. We will later do an enhancement by grouping in terms of eventSource of events.

**Review by krishna2020 — APPROVED** [review]

File: src/v0/destinations/tiktok_ads/util.js (hunk 1 above)
Comment (krishna2020) [inline src/v0/destinations/tiktok_ads/util.js:75]: Is camel case right standard for attributes ?

Comment (krishna2020) [reply src/v0/destinations/tiktok_ads/util.js:75]: Shouldn't it be properties.event_source ?
Ref: https://www.rudderstack.com/docs/event-spec/ecommerce-events-spec/browsing/
We follow snake case in e-commerce spec

Comment (ItsSudip) [reply src/v0/destinations/tiktok_ads/util.js:75]: For any attributes on RudderStack we generally do in camel case.

Comment (ItsSudip) [reply src/v0/destinations/tiktok_ads/util.js:75]: https://www.rudderstack.com/docs/destinations/streaming-destinations/tiktok-ads/cloud-mode/#property-mapping you can refer this mapping we are prioritizing anything in camelCase.

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Review by maheshkutty — APPROVED** [review]

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Review by koladilip — DISMISSED** [review]

File: src/v0/destinations/tiktok_ads/transformV2.js [hunk 3]
```diff
@@ -54,6 +61,18 @@ const getTrackResponsePayload = (message, destConfig, event, setDefaultForConten
...
+  // if contents is present then we need to add contents_ids and num_items to the payload
+  if (payload.properties?.contents?.length > 0) {
+    const contentIds = payload.properties.contents
+      .map((content) => content.content_id)
+      .filter(Boolean);
+
+    if (contentIds.length > 0) {
+      payload.properties.contents_ids = contentIds;
+    }
+    payload.properties.num_items = payload.properties.contents.length;
```
Comment (krishna2020) [inline src/v0/destinations/tiktok_ads/transformV2.js:73]: should it be contentIds length instead ?

File: src/v0/destinations/tiktok_ads/transformV2.js [hunk 4]
```diff
@@ -679,66 +719,71 @@ Returns
...
   let metadata = [];
-  const { destination } = eventsChunk[0];
+  const { destination } = processedEvents[0];
   const { pixelCode } = destination.Config;
-  eventsChunk.forEach((event) => {
+  processedEvents.forEach((event) => {
     const eventData = event.message[0]?.body.JSON.data;
     // eslint-disable-next-line unicorn/consistent-destructuring
-    if (Array.isArray(eventData) && eventData?.length > config.maxBatchSizeV2 - data.length) {
+    if (eventData.length + data.length > config.maxBatchSizeV2) {
```
Comment (krishna2020) [inline src/v0/destinations/tiktok_ads/transformV2.js:732]:
```suggestion
    if (eventData?.length + data?.length > config.maxBatchSizeV2) {
```

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)

**Review by krishna2020 — APPROVED** [review]

**Comment by devops-github-rudderstack** [comment]
Allure Test reports for this run are available at:
- Allure Report: [View Report](https://rudderlabs.github.io/test-reports/rudder-transformer/4497/index.html)
